src/app/codegen-data-model-provider/EmberDataBuffer.cpp

```diff
@@ -0,0 +1,352 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/codegen-data-model-provider/EmberDataBuffer.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/attribute-storage-null-handling.h>
+#include <app/util/odd-sized-integers.h>
+#include <lib/core/CHIPError.h>
+#include <lib/core/TLVTypes.h>
+#include <lib/support/CodeUtils.h>
+#include <protocols/interaction_model/Constants.h>
+#include <protocols/interaction_model/StatusCode.h>
+
+#include <limits>
+
+namespace chip {
+namespace app {
+namespace Ember {
+
+namespace {
+
+/// Maximum length of a string, inclusive
+///
+/// the max size value (0xFF and 0xFFFF) is reserved for NULL representation so
+/// it is not available
+constexpr uint32_t MaxLength(EmberAttributeBuffer::PascalStringType s)
+{
+    if (s == EmberAttributeBuffer::PascalStringType::kShort)
+    {
+        return std::numeric_limits<uint8_t>::max() - 1;
+    }
+    // EmberAttributeBuffer::PascalStringType::kLong:
+    return std::numeric_limits<uint16_t>::max() - 1;
+}
+
+struct UnsignedDecodeInfo
+{
+    unsigned byteCount;
+    uint64_t maxValue;
+};
+
+constexpr UnsignedDecodeInfo GetUnsignedDecodeInfo(EmberAfAttributeType type)
+{
+
+    switch (type)
+    {
+    case ZCL_INT8U_ATTRIBUTE_TYPE: // Unsigned 8-bit integer
+        return UnsignedDecodeInfo{ 1, NumericLimits::MaxUnsignedValue(1) };
```

Comment:
I would suggest this would be more readable if UnsignedDecodeInfo had a constepr constructor that took the byte count and did the right thing with it.